### PR TITLE
A configurable training pipeline.

### PR DESCRIPTION
**Pull request by tikurahul**
* We start off with a list of shallow records, because it makes
  things easier with a train_test_split.
* The rest of the pipeline is lazy.
* The `pipeline` module is typed.

### CODE REVIEW

File: donkeycar/pipeline/types.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+from typing_extensions import TypedDict
+
+TubRecordDict = TypedDict(
+    'TubRecordDict',
+    {
+        'cam/image_array': str,
+        'user/angle': float,
+        'user/throttle': float,
+        'user/mode': str,
+        'imu/acl_x': Optional[float],
```
Comment (DocGarbanzo) [inline donkeycar/pipeline/types.py:15]: We should use two 3-D arrays here, one for acceleration one for gyro.

**Review by DocGarbanzo — COMMENTED** [review]
Looks good to me.

File: donkeycar/pipeline/types.py (hunk 1 above)
Comment (tikurahul) [reply donkeycar/pipeline/types.py:17]: I am keeping the representation of the underlying dictionary here the same as what is actually stored.

Comment (tikurahul) [reply donkeycar/pipeline/types.py:17]: I am keeping it the same as what is actually stored.

**Comment by tikurahul** [comment]
This works as expected. I think we are pretty close to an eventual answer. 
Linear h5_pred

I realized that both `tf.data` and the old `Sequence` based API are actually have a slightly orthogonal API.
All our pipelines are `Iterator`s. Where as both `tf.Data` and `Sequence` expect a list of shallow rows to be an indexable `List` like type. Notice they dispatch a call to `__get_item__(self, index)`.

I think our models need to return a pipeline of `numpy.ndarray` producers. So imagine a type:

```python
class ToNumpy(object):
    def __init__(self) -> None:
         # Does nothing
         pass
     
    def to_nd_array(self):
         raise NotImplemented
```

Once we have this type, if we had a `SizedIterator[Tuple[ToNumpy, ToNumpy]]` then we can easily convert it to a `tf.Data` or a `Sequence` type. We are essentially one step away from a super nice API.

File: donkeycar/pipeline/sequence.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+
+class BatchedIterator(SizedIterator[List[R]]):
+    def __init__(self, pipeline: SizedIterator[R], batch_size: int) -> None:
+        self.pipeline = pipeline
+        self.batch_size = batch_size
+        self.consumed = 0
+
+    def __len__(self):
+        return math.ceil(len(self.pipeline) / self.batch_size)
```
Comment (DocGarbanzo) [inline donkeycar/pipeline/sequence.py:138]: Good to use `ceil` because before we were neglecting the the last non-completely filled batch.

File: donkeycar/pipeline/types.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    def image(self, cached=True, normalize=False) -> np.ndarray:
+        if not self._image:
+            image_path = self.underlying['cam/image_array']
+            full_path = os.path.join(self.base_path, 'images', image_path)
+            _image = load_image_arr(full_path, cfg=self.config)
+            if normalize:
+                _image = normalize_image(_image)
+            if cached:
+                self._image = _image
```
Comment (DocGarbanzo) [inline donkeycar/pipeline/types.py:43]: I think here we need to save the uint8 image in order to save memory, as the float array is w/ double precision, i.e. 8 times the size - and do the `normalize_image` on the fly.

File: setup.py [hunk 4]
```diff
@@ -46,7 +46,8 @@ def package_files(directory, strip_leading):
           'PrettyTable',
           'paho-mqtt',
           "simple_pid",
-          'progress'
+          'progress',
+          'typing_extensions',
```
Comment (DocGarbanzo) [inline setup.py:50]: We need to update the yaml files as well.

**Review by DocGarbanzo — APPROVED** [review]
Minor comments.

**Comment by DocGarbanzo** [comment]
> This works as expected. I think we are pretty close to an eventual answer.
> Linear h5_pred
> 
> I realized that both `tf.data` and the old `Sequence` based API are actually have a slightly orthogonal API.
> All our pipelines are `Iterator`s. Where as both `tf.Data` and `Sequence` expect a list of shallow rows to be an indexable `List` like type. Notice they dispatch a call to `__get_item__(self, index)`.
> 
> I think our models need to return a pipeline of `numpy.ndarray` producers. So imagine a type:
> 
> ```python
> class ToNumpy(object):
>     def __init__(self) -> None:
>          # Does nothing
>          pass
>      
>     def to_nd_array(self):
>          raise NotImplemented
> ```
> 
> Once we have this type, if we had a `SizedIterator[Tuple[ToNumpy, ToNumpy]]` then we can easily convert it to a `tf.Data` or a `Sequence` type. We are essentially one step away from a super nice API.

Yeah `tf.Sequence` is just a sequence requiring `__get_item__(self, index)` - but  you can create `tf.Data` also from a generator (or iterator, probably too), as it has a method `from_generator()`.
